### ISSUE TITLE
Modules: can't use variable in source

### ISSUE THREAD

**Issue by Alexhha** [hashicorp/terraform#18725]
### Terraform Version
```
$ terraform -v
Terraform v0.11.8
```

### Terraform Configuration Files
```hcl
// main.tf
provider "google" {
    credentials = "${file("~/.gce/gce-key.json")}"
    project     = "${var.project}"
    region      = "${var.region}"
    version     = "~> 1.12"
}

provider "random" {
    version = "~> 1.3"
}

provider "template" {
    version = "~> 1.0"
}

locals {
    app_module_source = "${var.app_ha_enabled == "true" ? "modules/app_ha" : "modules/app_standalone"}"
}

module "app" {
    source         = "${local.app_module_source}"
    region         = "europe-west1"
    project        = "${var.project}"
    gce_subnetwork = "${var.gce_subnetwork}"
}
```
```hcl
//variables.tf 
variable "project" {
    type = "string"
    default = "test"
}

variable "gce_subnetwork" {
    type = "string"
    default = "test"
}

variable "gce_zone" {
    type = "string"
    default = "europe-west1-b"
}

variable "region" {
    type = "string"
    default = "europe-west1"
}

variable "app_ha_enabled" {
    type = "string"
    default = "true"
}
```

### Debug Output
```
$ TF_LOG=trace terraform init
2018/08/22 17:03:12 [INFO] Terraform version: 0.11.8  7a5c1d221ac209bbac66947c369815cd9ca70ed5
2018/08/22 17:03:12 [INFO] Go runtime version: go1.10.1
2018/08/22 17:03:12 [INFO] CLI args: []string{"/usr/local/bin/terraform", "init"}
2018/08/22 17:03:12 [DEBUG] Attempting to open CLI config file: /home/alex/.terraformrc
2018/08/22 17:03:12 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2018/08/22 17:03:12 [INFO] CLI command args: []string{"init"}
2018/08/22 17:03:12 [DEBUG] command: loading backend config file: /home/alex/sandbox/terraform/test
Initializing modules...
2018/08/22 17:03:12 [TRACE] module source: "${local.app_module_source}"
- module.app
2018/08/22 17:03:12 [TRACE] detected module source "file:///home/alex/sandbox/terraform/test/${local.app_module_source}"
2018/08/22 17:03:12 [DEBUG] fetching module from file:///home/alex/sandbox/terraform/test/${local.app_module_source}
2018/08/22 17:03:12 [DEBUG] fetching "file:///home/alex/sandbox/terraform/test/${local.app_module_source}" with key "1.app;${local.app_module_source}"
2018/08/22 17:03:12 [DEBUG] plugin: waiting for all plugin processes to complete...
  Getting source "${local.app_module_source}"

Error downloading modules: Error loading modules:
error downloading 'file:///home/alex/sandbox/terraform/test/${local.app_module_source}':
source path error: stat /home/alex/sandbox/terraform/test/${local.app_module_source}:
no such file or directory
```
### Expected Behavior
By default it should include module /home/alex/sandbox/terraform/test/modules/app_ha. I was unable to found any restrictions for field source. Did I miss something? Also I have tried to use various combination of variable interpolation (including builtin functions) without success.

### Actual Behavior
Get the following error
```
source path error: stat /home/alex/sandbox/terraform/test/${local.app_module_source}:
no such file or directory
```

### Steps to Reproduce
```
$ terraform init
```

**Comment by apparentlymart** [comment]
Hi @Alexhha! Sorry this didn't work as expected.

The module tree is not dynamic, which is to say that it must be known during Terraform's initial analysis pass in order to construct the graph. Therefore expression evaluation is not permitted there, similarly to how in a static programming language you generally can't dynamically import other packages decided at runtime.

In current Terraform releases, the usual pattern here is to pass the relevant setting down into the child module and have the resources within disable themselves (e.g. by setting `count = 0`, or whatever else makes sense in each case), making the module tree static while the resource instances within it dynamic.

In a future version of Terraform we plan to implement support for using the `count` argument within module blocks, as discussed in #953. That'll provide a more convenient way to dynamically disable an entire module, so it'll still exist in the static module tree but Terraform will skip over it during evaluation once it's able to determine the value of `count`:

```hcl
# Not yet implemented, and details may change before release

module "app_ha" {
  count = var.app_ha_enabled ? 1 : 0

  source         = "./modules/app_ha"
  region         = "europe-west1"
  project        = var.project
  gce_subnetwork = var.gce_subnetwork
}

module "app_standalone" {
  count = var.app_ha_enabled ? 0 : 1

  source         = "./modules/app_standalone"
  region         = "europe-west1"
  project        = var.project
  gce_subnetwork = var.gce_subnetwork
}
```

In this particular case it may be _better_ to pass the variable down into the module and handle it there, if the differences between these two modules are in the details rather than in the overall structure, but at least this will give some additional options for implementing this sort of configurable behavior.

**Comment by Alexhha** [comment]
Will this feature (count) available in 0.12?

**Comment by apparentlymart** [comment]
Groundwork is being laid in the 0.12 development branch, but this particular feature will not land until a later release due to its relative complexity compared to the other changes. For more information, see [the announcement article describing this set of features](https://www.hashicorp.com/blog/hashicorp-terraform-0-12-preview-for-and-for-each).

**Comment by Alexhha** [comment]
Got it. And what about my use-case. When you implement count will be available to use variables in module source?

**Comment by apparentlymart** [comment]
It will never be possible to use variables in the module source. That is unfortunately not technically possible within the design constraints of the language, for the reasons I described in my previous comment.

Once implemented, #953 will be the way to meet this use-case, using a pattern like my example above (though potentially with additional use of variables to reduce the duplication). With that said, I'm going to close this one just to consolidate the discussion over there.
